src/main/java/org/elasticsearch/script/expression/FieldDataValueSource.java

```diff
@@ -49,13 +48,24 @@ public FunctionValues getValues(Map context, LeafReaderContext leaf) throws IOEx
     }
 
     @Override
-    public boolean equals(Object other) {
-        return fieldData.equals(other);
+    public boolean equals(Object o) {
```

Comment:
You can't assume that the incoming object to be compared against won't be null.  You can only assume certain members of the incoming object (assuming it's not null) won't be null.